### PR TITLE
Replace Vec with LiteMap in locale canonicalizer

### PR DESCRIPTION
**Pull request by pandusonu2**
Part 1 for #1034

### CODE REVIEW

File: components/locale_canonicalizer/src/locale_canonicalizer.rs
Comment (Manishearth) [inline components/locale_canonicalizer/src/locale_canonicalizer.rs:216]: nit: spelling

File: components/locale_canonicalizer/Cargo.toml
Comment (Manishearth) [inline components/locale_canonicalizer/Cargo.toml:36]: issue: The feature should not be enabled here, it should be enabled in `provider_serde` below as `"litemap/serde_serialize"`

**Review by Manishearth — CHANGES_REQUESTED** [review]
overall lgtm, minor issues

File: components/locale_canonicalizer/Cargo.toml
Comment (pandusonu2) [reply components/locale_canonicalizer/Cargo.toml]: Done, but may I know what the difference is?

**Review by Manishearth — APPROVED** [review]
r+ but i'd wait for shane to have a look too

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
LGTM!  Before merging, we should wait for @zbraniecki or @dminor to review since they are the component owners.  However, in the mean time, you can get started on part 2 of the ticket.  Thank you!

**Comment by pandusonu2** [comment]
@dminor / @zbraniecki PR review ping

File: components/locale_canonicalizer/src/locale_canonicalizer.rs
Comment (zbraniecki) [inline components/locale_canonicalizer/src/locale_canonicalizer.rs:217]: `nit (n-b)`: I'd rather inline as:

```rust
        if let Some(language_identifier) = $table.get(&&$key) {
```

**Review by zbraniecki — APPROVED** [review]
lgtm.
did you get a chance to compare performance on `cargo criterion` between main and your patch?

I hope no significant changes will be noticed, but it would be good to verify that before landing.

**Comment by pandusonu2** [comment]
> lgtm. did you get a chance to compare performance on `cargo criterion` between main and your patch?
> 
> I hope no significant changes will be noticed, but it would be good to verify that before landing.

Can you help me understand how to use this command? I'm getting "no such command" and if I try to cargo install: "there is nothing to install in `criterion v0.3.5`, because it has no binaries"

**Comment by zbraniecki** [comment]
It's a bit confusing, but you need to install this https://crates.io/crates/cargo-criterion

Technically you can just use cargo bench command without it, but this improves stdev

**Comment by pandusonu2** [comment]
> It's a bit confusing, but you need to install this https://crates.io/crates/cargo-criterion
> 
> Technically you can just use cargo bench command without it, but this improves stdev

Well, I still dont really understand how to read it, but heres the input. I ran in on main, followed by rebase and running on the patch:

```Line Break/Latin1/En    time:   [5.2817 us 5.2963 us 5.3152 us]                                  
                        change: [-11.619% -9.8022% -7.8190%] (p = 0.00 < 0.05)
                        Performance has improved.
Line Break/Latin1/En CSS                                                                             
                        time:   [9.7197 us 9.7932 us 9.8808 us]
                        change: [-5.3198% -3.4734% -1.7350%] (p = 0.00 < 0.05)
                        Performance has improved.

Line Break/UTF8/En      time:   [13.783 us 13.897 us 14.034 us]                                
                        change: [+4.9488% +7.1454% +9.5689%] (p = 0.00 < 0.05)
                        Performance has regressed.
Benchmarking Line Break/UTF8/Th: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.4s, enable flat sampling, or reduce sample count to 50.
Line Break/UTF8/Th      time:   [1.8208 ms 1.8302 ms 1.8408 ms]                                
                        change: [-4.4038% -1.9874% +0.2680%] (p = 0.11 > 0.05)
                        No change in performance detected.

Line Break/UTF16/En     time:   [12.351 us 12.485 us 12.639 us]                                 
                        change: [+0.8102% +3.5446% +5.8378%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Line Break/UTF16/En CSS time:   [12.530 us 12.639 us 12.769 us]                                     
                        change: [-6.0991% -4.5336% -3.0102%] (p = 0.00 < 0.05)
                        Performance has improved.
Line Break/UTF16/Th     time:   [12.592 us 12.703 us 12.836 us]                                 
                        change: [-4.1669% -2.5160% -0.7591%] (p = 0.00 < 0.05)
                        Change within noise threshold.

thread 'main' panicked at 'Failed to get fixture.: Os { code: 2, kind: NotFound, message: "No such file or directory" }', components/datetime/benches/datetime.rs:23:47
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: Criterion.rs benchmark target datetime exited with exit status: 101
```

**Comment by sffc** [comment]
You should run `cargo criterion` from the crate, not the repo root.

**Comment by pandusonu2** [comment]
> You should run `cargo criterion` from the crate, not the repo root.

My bad, this makes much more sense 

```
Finished bench [optimized + debuginfo] target(s) in 28.30s
canonicalize/create     time:   [13.403 us 13.449 us 13.503 us]                                 
                        change: [-2.3628% -1.3492% -0.3147%] (p = 0.01 < 0.05)
                        Change within noise threshold.
canonicalize/create+canonicalize                                                                             
                        time:   [33.460 us 33.639 us 33.939 us]
                        change: [-1.6194% -0.8808% +0.1136%] (p = 0.04 < 0.05)
                        Change within noise threshold.

canonicalize-noop/create                                                                             
                        time:   [2.3426 us 2.3474 us 2.3532 us]
                        change: [-1.7158% -0.5034% +0.6529%] (p = 0.43 > 0.05)
                        No change in performance detected.
canonicalize-noop/create+canonicalize                                                                             
                        time:   [5.1293 us 5.1509 us 5.1756 us]
                        change: [-0.3326% +0.4591% +1.2314%] (p = 0.29 > 0.05)
                        No change in performance detected.

likelysubtags/maximize  time:   [2.2002 us 2.2099 us 2.2209 us]                                    
                        change: [-6.5665% -4.2107% -2.1785%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Vec binary search -> Map should definitely improve the time, and so it shows

File: components/locale_canonicalizer/src/locale_canonicalizer.rs
Comment (pandusonu2) [reply components/locale_canonicalizer/src/locale_canonicalizer.rs]: Done

**Comment by Manishearth** [comment]
Note that LiteMap also works off of binary search, so I'm actually surprised there's a difference. I bet it's a cache thing: LiteMap is stored as two separate vectors, so the _search_ part can efficiently search a smaller vector that will often fit in cache.

(Rust's HashMap uses a similar optimization: it has a single buffer but the buffer is split into an occupancy bit vector, a key vector, and a value vector, with the occupancy vector being _very_ efficient for robin hood hashing)

**Comment by zbraniecki** [comment]
> Vec binary search -> Map should definitely improve the time, and so it shows

This is not the case. As Manish pointed out, this map uses binary search under the hood.

My read of the results you shared is that they don't indicate regression, which is all I was hoping to see. There may be some significance to the `maximize` result since p95 looks decent, but either way, I think we're good to merge this from the perf standpoint.

**Comment by pandusonu2** [comment]
> > Vec binary search -> Map should definitely improve the time, and so it shows
> 
> This is not the case. As Manish pointed out, this map uses binary search under the hood.
> 
> My read of the results you shared is that they don't indicate regression, which is all I was hoping to see. There may be some significance to the `maximize` result since p95 looks decent, but either way, I think we're good to merge this from the perf standpoint.

I dont really have merge permissions I think (from prev PR exp)..... nor is this PR approved 😅  Please do the needful

**Review by zbraniecki — APPROVED** [review]

**Pull request merged**
